src/main/utility/synchronization/count_down_latch.rs

```diff
@@ -0,0 +1,376 @@
+#[derive(Debug)]
+pub struct LatchCounter {
+    inner: Arc<LatchInner>,
+    /// This counter's generation.
+    generation: usize,
+}
+
+/// A latch waiter.
+///
+/// If a waiter is cloned, it will inherit the waiter's state for the current generation. For
+/// example if a waiter is cloned after it has already waited, then the new waiter will also be
+/// treated as if it had already waited in the current generation. If a waiter is cloned before it
+/// has waited, then the new waiter will also need to wait in the current generation.
+#[derive(Debug)]
+pub struct LatchWaiter {
+    inner: Arc<LatchInner>,
+    /// This waiter's generation.
+    generation: usize,
+}
+
+#[derive(Debug)]
+struct LatchInner {
+    lock: Mutex<LatchState>,
+    cond: Condvar,
+}
+
+#[derive(Debug)]
+struct LatchState {
+    /// The current latch "round".
+    generation: usize,
+    /// Number if counters remaining.
+    counters: usize,
+    /// Number of waiters remaining.
+    waiters: usize,
+    /// Total number of counters.
+    total_counters: usize,
+    /// Total number of waiters.
+    total_waiters: usize,
+}
+
+/// Build a latch counter and waiter. The counter and waiter can be cloned to create new counters
+/// and waiters.
+pub fn count_down_latch() -> (LatchCounter, LatchWaiter) {
+    let inner = Arc::new(LatchInner {
+        lock: Mutex::new(LatchState {
+            generation: 0,
+            counters: 1,
+            waiters: 1,
+            total_counters: 1,
+            total_waiters: 1,
+        }),
+        cond: Condvar::new(),
+    });
+
+    let counter = LatchCounter {
+        inner: Arc::clone(&inner),
+        generation: 0,
+    };
+
+    let waiter = LatchWaiter {
+        inner,
+        generation: 0,
+    };
+
+    (counter, waiter)
+}
+
+impl LatchState {
+    pub fn next_generation(&mut self) {
+        debug_assert_eq!(self.counters, 0);
+        debug_assert_eq!(self.waiters, 0);
+        self.counters = self.total_counters;
+        self.waiters = self.total_waiters;
+        self.generation = self.generation.wrapping_add(1);
+    }
+}
+
+impl LatchCounter {
+    /// Decrement the latch count and wake the waiters if the count reaches 0. This must not be
+    /// called more than once per generation (must not be called again until all of the waiters have
+    /// returned from their [`wait`] calls), otherwise it will panic.
+    pub fn count_down(&mut self) {
+        let counters;
+        {
+            let mut lock = self.inner.lock.lock().unwrap();
+
+            if self.generation != lock.generation {
+                let latch_gen = lock.generation;
+                std::mem::drop(lock);
+                panic!(
+                    "Counter generation does not match latch generation ({} != {})",
+                    self.generation, latch_gen
+                );
+            }
+
+            lock.counters = lock.counters.checked_sub(1).unwrap();
+            self.generation = self.generation.wrapping_add(1);
```

Comment:
Can we do this increment while we're not holding the lock?

Reply:
Moved this line to the end.